File: ask-llm.py
#!/usr/bin/env python3

import asyncio
import os
import json
import urllib.request
import time

LLM_API_BASE_URL = os.environ.get("LLM_API_BASE_URL", "https://api.openai.com/v1")
LLM_API_KEY = os.environ.get("LLM_API_KEY") or os.environ.get("OPENAI_API_KEY")
LLM_CHAT_MODEL = os.environ.get("LLM_CHAT_MODEL")
LLM_STREAMING = os.environ.get("LLM_STREAMING", "yes") != "no"

LLM_DEBUG = os.environ.get("LLM_DEBUG")


async def chat(messages, handler=None):
    url = f"{LLM_API_BASE_URL}/chat/completions"
    auth_header = f"Bearer {LLM_API_KEY}" if LLM_API_KEY else None
    headers = {
        "Content-Type": "application/json",
        "User-Agent": "python-requests/2.31.0",
    }
    if auth_header:
        headers["Authorization"] = auth_header

    model = LLM_CHAT_MODEL or "gpt-4o-mini"
    stop = ["<|im_end|>", "<|end|>", "<|eot_id|>"]
    max_tokens = 200
    temperature = 0
    stream = LLM_STREAMING and callable(handler)

    body = {
        "messages": messages,
        "model": model,
        "stop": stop,
        "max_tokens": max_tokens,
        "temperature": temperature,
        "stream": stream,
    }
    json_body = json.dumps(body).encode("utf-8")
    request = urllib.request.Request(
        url, data=json_body, headers=headers, method="POST"
    )
    response = urllib.request.urlopen(request)

    if not stream:
        if response.status != 200:
            raise Exception(f"HTTP error: {response.status} {response.reason}")
        data = json.loads(response.read().decode("utf-8"))
        choices = data["choices"]
        first = choices[0]
        message = first["message"]
        content = message["content"]
        full_answer = content.strip()
        if handler:
            handler(full_answer)
        return full_answer
    else:

        def parse(line):
            partial = None
            prefix = line[:6]
            if prefix == "data: ":
                payload = line[6:]
                try:
                    choices = json.loads(payload)["choices"]
                    choice = choices[0]
                    delta = choice.get("delta", {})
                    partial = delta.get("content", "")
                except Exception as e:
                    pass
            return partial

        finished = False
        buffer = []
        answer = ""
        while not finished:
            raw_bytes = response.read(8)
            if not raw_bytes:
                break
            buffer.append(raw_bytes)
            lines = b"".join(buffer).decode("utf-8").splitlines(True)
            full_answer = ""
            for line in lines:
                if len(line) > 0:
                    if line[0] == ":":
                        continue
                    if line == "data: [DONE]":
                        finished = True
                        break
                    elif line:
                        partial = parse(line)
                        if partial is not None:
                            if len(full_answer) == 0:
                                full_answer = partial.strip()
                            else:
                                full_answer += partial
            if handler:
                handler(full_answer.replace(answer, ""))
            answer = full_answer

        return answer


SYSTEM_PROMPT = "Answer the question politely and concisely."


async def main():
    print(f"Using LLM at {LLM_API_BASE_URL}.")
    print("Press Ctrl+D to exit.")
    print()

    messages = [{"role": "system", "content": SYSTEM_PROMPT}]

    while True:
        try:
            question = input(">> ")
        except EOFError:
            break

        messages.append({"role": "user", "content": question})
        start = time.time()
        stream = lambda partial: print(partial, end="", flush=True)
        answer = await chat(messages, stream)
        messages.append({"role": "assistant", "content": answer})
        print()
        elapsed = time.time() - start
        if LLM_DEBUG:
            print(f"[{round(elapsed * 1000)} ms]")
        print()


asyncio.run(main())

File: ask-llm.clj
#!/usr/bin/env bb

(require '[clojure.string :as str]
         '[babashka.http-client :as http]
         '[cheshire.core :as json])

(def LLM-API-BASE-URL (or (System/getenv "LLM_API_BASE_URL") "https://api.openai.com/v1"))
(def LLM-API-KEY (or (System/getenv "LLM_API_KEY") (System/getenv "OPENAI_API_KEY")))
(def LLM-CHAT-MODEL (System/getenv "LLM_CHAT_MODEL"))
(def LLM-STREAMING (not= "no" (System/getenv "LLM_STREAMING")))

(def LLM-DEBUG (System/getenv "LLM_DEBUG"))

(defn http-json-headers [bearer]
  (if bearer {:content-type "application/json"
              :authorization (str "Bearer " bearer)}
      {:content-type "application/json"}))

(def LLM-CHAT-URL (str LLM-API-BASE-URL "/chat/completions"))

(def SYSTEM-PROMPT "Answer the question politely and concisely.")

(defonce llm-messages (atom [{:role "system"
                              :content SYSTEM-PROMPT}]))

(defn add-message! [role content]
  (swap! llm-messages conj {:role role :content content}))

(defn make-reader [response]
  (java.io.BufferedReader. (java.io.InputStreamReader. (:body response))))

(defn json-parse [str]
  (json/parse-string str true))

(defn parse-line [line handler]
  (try
    (some-> line json-parse
            :choices first :delta :content
            (#(do (when handler (handler %))
                  (str %))))
    (catch Exception e nil)))

(defn decode-stream [response handler]
  (with-open [reader (make-reader response)]
    (loop [answer ""]
      (if-let [line (.readLine reader)]
        (let [trimmed-line (str/trim line)]
          (cond
            (str/blank? trimmed-line) (recur answer)
            (str/starts-with? trimmed-line "data: ")
            (let [new-answer (parse-line (str/trim (subs trimmed-line 6)) handler)]
              (recur (or new-answer answer)))
            :else (recur answer)))
        answer))))

(defn chat [messages handler]
  (let [stream (and LLM-STREAMING (some? handler))
        payload {:messages messages
                 :model (or LLM-CHAT-MODEL "gpt-4o-mini")
                 :stop ["<|im_end|>" "<|end|>" "<|eot_id|>"]
                 :max_tokens 200
                 :temperature 0
                 :stream stream}
        options {:headers (http-json-headers LLM-API-KEY)
                 :body (json/encode payload)}
        options (if stream (assoc options :as :stream) options)
        response (http/post LLM-CHAT-URL options)]
    (if stream
      (decode-stream response handler)
      (let [body (-> response :body json-parse)
            answer (-> body :choices first :message :content str/trim)]
        (when handler (handler answer))
        answer))))

(defn print-stdout [str]
  (print str)
  (flush))

(defn ask-llm [question]
  (add-message! "user" question)
  (let [answer (chat @llm-messages print-stdout)]
    (add-message! "assistant" answer)
    (println)
    answer))

(defmacro measure-time [f]
  `(let [start-time# (System/currentTimeMillis)
         result# ~f
         end-time# (System/currentTimeMillis)]
     (when LLM-DEBUG (println (str "[" (- end-time# start-time#) " ms]")))
     result#))

(defn qa []
  (loop []
    (print ">> ")
    (flush)
    (let [question (read-line)]
      (when question
        (-> question ask-llm measure-time)
        (println)
        (flush)
        (recur)))))

(defn -main []
  (println "Using LLM at" LLM-API-BASE-URL)
  (println "Press Ctrl+D to exit.")
  (println)
  (qa))

(-main)
